storage/merkle/node.go

```diff
@@ -3,23 +3,110 @@
+type node interface {
+	// Hash computes recursively the hash of this respective sub trie.
+	// To simplify enforcing cryptographic security, we introduce the convention
+	// that hashing a nil node is an illegal operation, which panics.
+	Hash() []byte
+}
+
+// NodeTags encodes the type of node when hashing it. Required for cryptographic
+// safety to prevent collision attacks (replacing a full node with a short node
+// that has identical hash would otherwise be possible).
+// Values are intended to be global constants and must be unique.
+var (
+	leafNodeTag  = []byte{0}
+	fullNodeTag  = []byte{1}
+	shortNodeTag = []byte{2}
+)
+
+/* ******************************* Short Node ******************************* */
+// Per convention a Short node has always _one child_, which is either
+// a full node or a leaf.
 
 type short struct {
+	count int    // holds the count of bits in the path
 	path  []byte // holds the common path to the next node
-	count uint   // holds the count of bits in the path
-	child node   // holds the child after the common path
+	child node   // holds the child after the common path; never nil
 }
 
+var _ node = &short{}
+
+func (n *short) Hash() []byte {
+	c := serializedPathSegmentLength(n.count)
+	h, _ := blake2b.New256(shortNodeTag) // blake2b.New256(..) error for given MAC (verified in tests)
+	_, _ = h.Write(c[:])                 // blake2b.Write(..) never errors for _any_ input
+	_, _ = h.Write(n.path)               // blake2b.Write(..) never errors for _any_ input
+	_, _ = h.Write(n.child.Hash())       // blake2b.Write(..) never errors for _any_ input
+	return h.Sum(nil)
+}
+
+// serializedPathSegmentLength serializes the bitCount into two bytes using the following optimization:
+// A short node with zero path length is not part of our storage model. Therefore, we use the convention:
+//  * for path length l with 1 ≤ l ≤ 65535: we represent l as unsigned int with big-endian encoding
+//  * for l = 65536: we represent l as binary 00000000 00000000
+// This convention organically utilizes the natural occurring overflow and is therefore extremely
+// efficient. In summary, we are able to represent key length of up to 65536 bits, i.e. 8192 bytes.
+func serializedPathSegmentLength(bitCount int) [2]byte {
+	var byteCount [2]byte
+	byteCount[0] = byte(bitCount >> 8)
+	byteCount[1] = byte(bitCount)
+	return byteCount
+}
+
+/* ******************************** Full Node ******************************* */
+// Per convention a Full Node has always _two children_. Nil values not allowed.
+
 type full struct {
-	left  node // holds the left path node (bit 0)
-	right node // holds the right path node (bit 1)
+	left  node // holds the left path node (bit 0); never nil
+	right node // holds the right path node (bit 1); never nil
+}
+
+var _ node = &full{}
+
+func (n *full) Hash() []byte {
+	h, _ := blake2b.New256(fullNodeTag) // blake2b.New256(..) error for given MAC (verified in tests)
```

Comment:
Even if the lenght is fixed for full nodes It might be valuable to include `serializedPathSegmentLength` to be consistent with the short nodes.

Reply:
Sorry, not quite sure what you mean. Full nodes don't have a path segment at all (neither a path segment of variable length nor a path segment with fixed length). 

Each of the `fullNodeTag`, `leafNodeTag`, `shortNodeTag` are _unique and constant_ 1-byte identifiers to include the node type in the cryptographic hash: https://github.com/onflow/flow-go/blob/aa285f72ff6cbb8c601178e909eb9f86bccfe9a9/storage/merkle/node.go#L25-L29
Their purpose is to make collision attacks impossible, where an attacker pretends that a node is of a different type than it actually is. 

I probably misunderstood your suggestion @ramtinms. Some clarification / further explanation would be very helpful. Thanks.

Reply:
all good, everything looks fine now